Add --build-plan for 'cargo build'

mshal opened the pull request:
With 'cargo build --build-plan', cargo does not actually run any
commands, but instead prints out what it would have done in the form of
a JSON data structure.

Fixes #3815

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

rust-highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* Pull requests are usually filed against the master branch for this repo, but this one is against rust-1.21.0. Please double check that you specified the right target!

mshal commented:
I filed this against rust-1.21.0 since that is what mozilla-central builds with. I can try to rebase against master if that would help, but IIRC the master version of cargo wouldn't build m-c the last time I tried.

cyplo commented:
Heya ! Any updates on this one ? Would this apply to master cleanly this time ?

mshal commented:
I've rebased to the latest cargo master here: https://github.com/rust-lang/cargo/pull/5301

I also updated it slightly to make the output consistent when executed multiple times.
